sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    val leftGen = if (left.dataType.sameType(StringType)) {
+      left.genCode(ctx)
+    } else {
+      left.asInstanceOf[Collate].inputString.genCode(ctx)
+    }
+    val rightGen = if (right.dataType.sameType(StringType)) {
+      right.genCode(ctx)
+    } else {
+      right.asInstanceOf[Collate].inputString.genCode(ctx)
+    }
+    collator match {
+      case null =>
+        collationID match {
+          case 0 =>
+            // UCS_BASIC
+            defineCodeGen(ctx, ev, (c1, c2) => s"($c1).contains($c2)")
+          case 1 =>
+            // UCS_BASIC_LCASE
+            defineCodeGen(ctx, ev, (c1, c2) => s"($c1.toLowerCase()).contains($c2.toLowerCase())")
+          case _ =>
+            // Other special collations without ICU collator
+            val resultCode =
+              s"""
+                 |${leftGen.code}
+                 |${rightGen.code}
+                 |org.apache.spark.sql.catalyst.util.CollatorFactory.CollatorInfo collatorInfo =
+                 |$collatorFactory.getInfoForId(collID);
+                 |int lLen = ${leftGen.value}.toString().length();
+                 |int rLen = ${rightGen.value}.toString().length();
+                 |boolean result = false;
+                 |for (int i = 0; i <= lLen - rLen && !result; i++) {
+                 |  if (collatorInfo.comparator.compare(${leftGen.value}.substring(i, i + rLen),
+                 |  ${rightGen.value}) == 0) {
+                 |    result = true;
+                 |  }
+                 |}
+                 |""".stripMargin
+            ev.copy(code =
+              code"""
+                    |${leftGen.code}
+                    |${rightGen.code}
+                    |boolean ${ev.isNull} = false;
+                    |$resultCode
+                    |boolean ${ev.value} = result;
+                    |""".stripMargin)
+        }
+      case _ =>
+        val resultCode =
+          s"""
+             |org.apache.spark.sql.catalyst.util.CollatorFactory.CollatorInfo collatorInfo =
+             |$collatorFactory.getInfoForId(${collationID});
+             |com.ibm.icu.text.Collator collator = collatorInfo.collator;
+             |String l = ${leftGen.value}.toString();
+             |String r = ${rightGen.value}.toString();
+             |boolean result = false;
+             |for (int l_start = 0; l_start <= l.length() - r.length() && !result; l_start++) {
+             |  if (collator.compare(l.substring(l_start, l_start + r.length()), r) == 0) {
+             |    result = true;
+             |  }
+             |}
+             |""".stripMargin
+        ev.copy(code =
+          code"""
+                |${leftGen.code}
+                |${rightGen.code}
+                |boolean ${ev.isNull} = false;
+                |$resultCode
+                |boolean ${ev.value} = result;
+                |""".stripMargin)
+    }
```

Comment:
Note: this is the manual implementation for all collations with collator (collationID >=2).

(see above for corresponding Scala code)

Reply:
Update: collator is no longer exposed outside Collation, so this is now irrelevant since the new [PR](https://github.com/apache/spark/pull/45216/files)